### PR TITLE
Improve TensorRT GetCapability to Enable More Models

### PR DESCRIPTION
**Pull request by stevenlix**

### CODE REVIEW

**Comment by jywu-msft** [comment]
seems like windows tensorrt build fails

**Comment by jywu-msft** [comment]
if GetCapability() is now reliable, doesn't this mean we can re-enable many of the TRT disabled unit tests?
The ultimate validation would be if the TensorRT Execution provider exclusions in https://github.com/microsoft/onnxruntime/pull/802 can be removed (or kept at an absolute minimum)

**Comment by stevenlix** [comment]
Yes. I think most of the disabled unit tests could be removed. But if we keep those tests disabled people know what tests can't run on TensorRT, otherwise we just lose the tracking since those tests just fall back to other execution providers.

**Comment by jywu-msft** [comment]
We do not want to leave tests disabled. It was only done as a temporary workaround until GetCapability is fixed. 
The purpose of the unit tests is to ensure correctness.  Falling back via GetCapability is still a correct/valid result. If we leave the tests disabled, there's no validation that the GetCapability() changes in this PR are working as expected. And it prevents future regressions in GetCapability().
Have all unit tests pass (without disabling any tests) gives us confidence that TRT provider can handle real models without crashing.

> Yes. I think most of the disabled unit tests could be removed. But if we keep those tests disabled people know what tests can't run on TensorRT, otherwise we just lose the tracking since those tests just fall back to other execution providers.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h [hunk 1]
```diff
@@ -74,16 +71,23 @@ class TensorrtExecutionProvider : public IExecutionProvider {
...
+    max_batch_size_ = batch_size;
   }
 
   void SetMaxWorkspaceSize(const size_t workspace_size) {
-      max_workspace_size_ = workspace_size;
+    max_workspace_size_ = workspace_size;
   }
-    
+
+  std::unique_ptr<IndexedSubGraph> GetSubGraph(SubGraph_t graph_nodes_index, int& kernels_index,
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: do all these methods here need to be public? (GetSubGraph, GetSupportedLst etc.)

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 2]
```diff
@@ -45,16 +45,200 @@ TensorrtExecutionProvider::TensorrtExecutionProvider()
 
 TensorrtExecutionProvider::~TensorrtExecutionProvider() {}
 
+std::unique_ptr<IndexedSubGraph> TensorrtExecutionProvider::GetSubGraph(SubGraph_t graph_nodes_index, int& kernels_index, const onnxruntime::GraphViewer& graph) const {
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: some comments describing the algorithm and motivation for the new approach for implementing GetCapability would be nice.
maybe add some comments for GetSubGraph and GetSupportedList

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h [hunk 3]
```diff
@@ -74,16 +71,23 @@ class TensorrtExecutionProvider : public IExecutionProvider {
...
+                                               const onnxruntime::GraphViewer& graph) const;
+
+  SubGraphCollection_t GetSupportedList(SubGraphCollection_t supported_nodes_list, int iterations, const int& max_iterations,
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: why is max_iterations a reference?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h [hunk 4]
```diff
@@ -74,16 +71,23 @@ class TensorrtExecutionProvider : public IExecutionProvider {
...
+                                        bool& early_termination, const onnxruntime::GraphViewer& graph) const;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: early_termination seems to be an output parameter?
if so, it should follow the inputs in the method signature.
also it shouldn't be reference, should be pointer for outputs.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 5]
```diff
@@ -426,18 +530,25 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
-        // TODO: shape inference
         const auto& shape_size = output_shapes[i].size();
         output_tensors[output_index].ndim = shape_size;
         output_tensors[output_index].shape = new int64_t[shape_size];
         memcpy(output_tensors[output_index].shape, &output_shapes[i][0], sizeof(int64_t) * shape_size);
-        output_tensors[output_index].data = (*(trt_state->test_allocate_func))(trt_state->allocator, 64, sizeof(double) * batch_size * output_dim_sizes[i]);
 
-        CHECK_CUDA(cudaMemcpy(output_tensors[output_index].data, buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(float), cudaMemcpyDeviceToHost));
+        int output_size = batch_size * output_dim_sizes[i];
+        if (output_types[i] == TensorProto::FLOAT) {
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: add comments for what we're doing here.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 6]
```diff
@@ -426,18 +530,25 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
+          output_tensors[output_index].dtype = DType::TFloat32;
+          output_tensors[output_index].data = (*(trt_state->test_allocate_func))(trt_state->allocator, 32, output_size * sizeof(float));
+          CHECK_CUDA(cudaMemcpy(output_tensors[output_index].data, buffers[i + num_binding_inputs], output_size * sizeof(float), cudaMemcpyDeviceToHost));
+        } else if (output_types[i] == TensorProto::INT64) {
+          output_tensors[output_index].dtype = DType::TInt64;
+          output_tensors[output_index].data = (*(trt_state->test_allocate_func))(trt_state->allocator, 64, output_size * sizeof(int64_t));
+          CHECK_CUDA(cudaMemcpy(output_tensors[output_index].data, buffers[i + num_binding_inputs], output_size * sizeof(int), cudaMemcpyDeviceToHost));
+        } else {
+          ORT_THROW("Output type (%i) is not supported by TensorRT", output_types[i]);
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: better to return error status rather than throw exception?

File: onnxruntime/test/providers/cpu/math/gemm_test.cc [hunk 7]
```diff
@@ -25,7 +23,7 @@ TEST(GemmOpTest, GemmNoTrans) {
   test.AddOutput<float>("Y", {2, 3},
                         {11.0f, 11.0f, 11.0f,
                          -9.0f, -9.0f, -9.0f});
-  test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kTensorrtExecutionProvider});
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "");
```
Comment (jywu-msft) [inline onnxruntime/test/providers/cpu/math/gemm_test.cc]: shouldn't these just be test.Run()   , don't need the other args.

Comment (jywu-msft) [reply onnxruntime/test/providers/cpu/math/gemm_test.cc]: same comment for all the below test.Run() invocations

**Comment by jywu-msft** [comment]
I see TRTProvider exclusions were removed in gemm_test.cc
what about other tests? there were quite a few enabled in https://github.com/microsoft/onnxruntime/pull/802
will we be able to remove all the exclusions, or are there still some failures in GetCapability?

File: onnxruntime/test/providers/cpu/tensor/upsample_op_test.cc [hunk 8]
```diff
@@ -173,7 +174,7 @@ TEST(UpsampleOpTest, UpsampleOpNearest222XTest) {
   };
 
   test.AddOutput<float>("Y", {N*2, C, (int64_t)(H * scales[2]), (int64_t)(W * scales[3])}, Y);
-  test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kTensorrtExecutionProvider});//TensorRT parser: Assertion failed: scales[0] == 1 && scales[1] == 1
+  test.Run();//TensorRT parser: Assertion failed: scales[0] == 1 && scales[1] == 1
```
Comment (jywu-msft) [inline onnxruntime/test/providers/cpu/tensor/upsample_op_test.cc]: should these assertion failure comments be removed if we are re-enabling the TensorrtExecutionProvider for these tests?

Comment (jywu-msft) [reply onnxruntime/test/providers/cpu/tensor/upsample_op_test.cc]: same comment for other tests.

File: .gitmodules [hunk 9]
```diff
@@ -28,6 +28,7 @@
 [submodule "cmake/external/onnx-tensorrt"]
 	path = cmake/external/onnx-tensorrt
 	url = https://github.com/onnx/onnx-tensorrt.git
+        branch = v5.0
```
Comment (snnn) [inline .gitmodules]: Is this change needed ?

Comment (stevenlix) [reply .gitmodules]: onnx-tensorrt master branch is under big changes for new tensorrt release, so we point to a working branch before onnx-tensorrt changes are done.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc (hunk 2 above)
Comment (stevenlix) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: Add more comments for those two new functions

File: onnxruntime/test/providers/cpu/tensor/upsample_op_test.cc (hunk 8 above)
Comment (stevenlix) [reply onnxruntime/test/providers/cpu/tensor/upsample_op_test.cc]: removed the comments

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 10]
```diff
@@ -489,12 +489,18 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
 
-        // If output tensor type is INT64, TensorRT processes data as INT32 and the output will be converted to INT64.
+        int output_size = batch_size * output_dim_sizes[i];
         OrtValue* output_tensor = ort.KernelContext_GetOutput(context, output_index, output_shapes[i].data(), output_shapes[i].size());
-        if (output_types[i] == TensorProto::FLOAT) {        
-          CHECK_CUDA(cudaMemcpy(ort.GetTensorMutableData<float>(output_tensor), buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(float), cudaMemcpyDeviceToHost));
+        // If output tensor type is INT64, TensorRT processes data as INT32 and the output will be converted to INT64.
+        if (output_types[i] == TensorProto::FLOAT) {
+          CHECK_CUDA(cudaMemcpy(ort.GetTensorMutableData<float>(output_tensor), buffers[i + num_binding_inputs], output_size * sizeof(float), cudaMemcpyDeviceToHost));
         } else if (output_types[i] == TensorProto::INT64) {
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: shouldn't there be a TensorProto:INT32 case?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 11]
```diff
@@ -489,12 +489,18 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
         int output_index = output_indexes[i];
         output_shapes[i].insert(output_shapes[i].begin(), batch_size);
 
-        // If output tensor type is INT64, TensorRT processes data as INT32 and the output will be converted to INT64.
+        int output_size = batch_size * output_dim_sizes[i];
         OrtValue* output_tensor = ort.KernelContext_GetOutput(context, output_index, output_shapes[i].data(), output_shapes[i].size());
-        if (output_types[i] == TensorProto::FLOAT) {        
-          CHECK_CUDA(cudaMemcpy(ort.GetTensorMutableData<float>(output_tensor), buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(float), cudaMemcpyDeviceToHost));
+        // If output tensor type is INT64, TensorRT processes data as INT32 and the output will be converted to INT64.
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: move the comment to where the applicable code is.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc (hunk 10 above)
Comment (stevenlix) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: So far all the models have either FLOAT or INT64 outputs. I plan to add other data types that Nvidia claimed to support such as INT8, FLOAT16 and INT32 in another PR once I tested them in the models.

Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: doesn't seem safe to go by the handful of models you manually tested.
what about the case of intermediate outputs (for partitioned subgraphs)
it seems safer add the types that are supported by TensorRT and supported in ONNX.
otherwise, your else case "output type not supported by TensorRT" is not really true and may cause confusion.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 12]
```diff
@@ -489,12 +489,18 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
         } else if (output_types[i] == TensorProto::INT64) {
-          CHECK_CUDA(cudaMemcpy(ort.GetTensorMutableData<int64_t>(output_tensor), buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(int), cudaMemcpyDeviceToHost));
+          int* output = new int[output_size];
+          CHECK_CUDA(cudaMemcpy(output, buffers[i + num_binding_inputs], output_size * sizeof(int), cudaMemcpyDeviceToHost));
+          for (int j = 0; j < output_size; ++j) {
+            ort.GetTensorMutableData<int64_t>(output_tensor)[j] = output[j];
+          }
+          delete[] output;
         } else {
           Status(common::ONNXRUNTIME, common::FAIL, "Output type is not supported by TensorRT");
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: be clear and concise in error messages. it's not useful to say the output type is not supported without specifying which type it is.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 13]
```diff
@@ -462,22 +461,44 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
         const OrtValue* input_tensor = ort.KernelContext_GetInput(context, input_indexes[i]);
         auto tensor_info = ort.GetTensorTypeAndShape(input_tensor);
         const auto& tensor_shape = ort.GetTensorShape(tensor_info);
+        auto tensor_type = ort.GetTensorElementType(tensor_info);
         ort.ReleaseTensorTypeAndShapeInfo(tensor_info);
-        const float* input = ort.GetTensorData<float>(input_tensor);
 
         const int input_batch_size = tensor_shape[0];
         if (i > 0 && batch_size != input_batch_size) {
           ORT_THROW("Input batch size is inconsistent");
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: thought we should avoid throwing exception in compute function?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 14]
```diff
@@ -462,22 +461,44 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
+        } else if (tensor_type == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8) {
+          const int8_t* input = const_cast<int8_t*>(ort.GetTensorData<int8_t>(input_tensor));
+          CHECK_CUDA(cudaMalloc(&buffers[i], input_size * sizeof(int8_t)));
+          CHECK_CUDA(cudaMemcpy(buffers[i], input, input_size * sizeof(int8_t), cudaMemcpyHostToDevice));
+        } else if (tensor_type == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32) {
+          const int32_t* input = const_cast<int32_t*>(ort.GetTensorData<int32_t>(input_tensor));
+          CHECK_CUDA(cudaMalloc(&buffers[i], input_size * sizeof(int32_t)));
+          CHECK_CUDA(cudaMemcpy(buffers[i], input, input_size * sizeof(int32_t), cudaMemcpyHostToDevice));
+        } else {
+          Status(common::ONNXRUNTIME, common::FAIL, "Input tensor type " + std::to_string(tensor_type) + " is not supported.");
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: why allocate a Status() , doesn't seem used anywhere.

Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: same comment for other Status() allocations.

Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: looks like this compute_func() always returns 0, which doens't seem correct.
if there are errors, we need to return 1.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc (hunk 13 above)
Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: I see other places in Compile() function where it uses ORT_ENFORCE instead of returning a Status
please revisit.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 15]
```diff
@@ -462,22 +461,44 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
       for (int i = 0, end = num_binding_outputs; i < end; ++i) {
-        CHECK_CUDA(cudaMalloc(&buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(float)));
+        if (output_types[i] == ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT) {
+          CHECK_CUDA(cudaMalloc(&buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(float)));
+        } else if (output_types[i] == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8) {
+          CHECK_CUDA(cudaMalloc(&buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(int8_t)));
+        } else if (output_types[i] == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32 || output_types[i] == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64) {
+          CHECK_CUDA(cudaMalloc(&buffers[i + num_binding_inputs], batch_size * output_dim_sizes[i] * sizeof(int32_t)));
+        } else {
+          Status(common::ONNXRUNTIME, common::FAIL, "Output tensor type " + std::to_string(output_types[i]) + " is not supported.");
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: same comment as above for Status()

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 16]
```diff
@@ -491,18 +512,22 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
+          // If output tensor type is INT64, TensorRT processes data as INT32 and the output will be converted to INT64.
+          int* output = new int32_t[output_size];
+          CHECK_CUDA(cudaMemcpy(output, buffers[i + num_binding_inputs], output_size * sizeof(int32_t), cudaMemcpyDeviceToHost));
           for (int j = 0; j < output_size; ++j) {
             ort.GetTensorMutableData<int64_t>(output_tensor)[j] = output[j];
           }
           delete[] output;
         } else {
-          Status(common::ONNXRUNTIME, common::FAIL, "Output type is not supported by TensorRT");
+          Status(common::ONNXRUNTIME, common::FAIL, "Output tensor type " + std::to_string(output_types[i]) + " is not supported.");
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: same Status() comment

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 17]
```diff
@@ -383,8 +509,26 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<onnxruntime:
...
+        } else if (output_types[i] == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64) {
+          // If output tensor type is INT64, TensorRT processes data as INT32 and the output will be converted to INT64.
+          int* output = new int32_t[output_size];
+          CHECK_CUDA(cudaMemcpy(output, buffers[i + num_binding_inputs], output_size * sizeof(int32_t), cudaMemcpyDeviceToHost));
+          for (int j = 0; j < output_size; ++j) {
+            ort.GetTensorMutableData<int64_t>(output_tensor)[j] = output[j];
+          }
+          delete[] output;
+        } else {
+          return 1;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: better to return specific status code instead of 1 (same for other places)
https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/common/ml_status.h

**Comment by jywu-msft** [comment]
looks like you'll need to merge master to pick up https://github.com/microsoft/onnxruntime/pull/1097
 to resolve the nocontribops CI failure.
